core/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/AutoDateHistogramAggregator.java

```diff
@@ -0,0 +1,196 @@
+    private final DocValueFormat formatter;
+    private final Rounding[] roundings;
+    private int roundingIdx = 0;
+
+    private LongHash bucketOrds;
+    private int targetBuckets;
+    private MergingBucketsDeferringCollector deferringCollector;
+
+    AutoDateHistogramAggregator(String name, AggregatorFactories factories, int numBuckets, Rounding[] roundings,
+            @Nullable ValuesSource.Numeric valuesSource, DocValueFormat formatter, SearchContext aggregationContext, Aggregator parent,
+            List<PipelineAggregator> pipelineAggregators, Map<String, Object> metaData) throws IOException {
+
+        super(name, factories, aggregationContext, parent, pipelineAggregators, metaData);
+        this.targetBuckets = numBuckets;
+        this.valuesSource = valuesSource;
+        this.formatter = formatter;
+        this.roundings = roundings;
+
+        bucketOrds = new LongHash(1, aggregationContext.bigArrays());
+
+    }
+
+    @Override
+    public boolean needsScores() {
+        return (valuesSource != null && valuesSource.needsScores()) || super.needsScores();
+    }
+
+    @Override
+    protected boolean shouldDefer(Aggregator aggregator) {
+        return true;
+    }
+
+    @Override
+    public DeferringBucketCollector getDeferringCollector() {
+        deferringCollector = new MergingBucketsDeferringCollector(context);
+        return deferringCollector;
+    }
+
+    @Override
+    public LeafBucketCollector getLeafCollector(LeafReaderContext ctx,
+            final LeafBucketCollector sub) throws IOException {
+        if (valuesSource == null) {
+            return LeafBucketCollector.NO_OP_COLLECTOR;
+        }
+        final SortedNumericDocValues values = valuesSource.longValues(ctx);
+        return new LeafBucketCollectorBase(sub, values) {
+            @Override
+            public void collect(int doc, long bucket) throws IOException {
+                assert bucket == 0;
+                if (values.advanceExact(doc)) {
+                    final int valuesCount = values.docValueCount();
+
+                    long previousRounded = Long.MIN_VALUE;
+                    for (int i = 0; i < valuesCount; ++i) {
+                        long value = values.nextValue();
+                        long rounded = roundings[roundingIdx].round(value);
+                        assert rounded >= previousRounded;
+                        if (rounded == previousRounded) {
+                            continue;
+                        }
+                        long bucketOrd = bucketOrds.add(rounded);
+                        if (bucketOrd < 0) { // already seen
+                            bucketOrd = -1 - bucketOrd;
+                            collectExistingBucket(sub, doc, bucketOrd);
+                        } else {
+                            collectBucket(sub, doc, bucketOrd);
+                            while (bucketOrds.size() > targetBuckets) {
+                                increaseRounding();
```

Comment:
this operation does two things:
 - compute the merge map (cheap since the number of buckets is small)
 - recompute buffered buckets (potentially expensive since the number of docs is not bounded)

Yet the merge map alone is enough to know how many buckets we will have in the end, so I think it should do something like this instead. I'm unsure how often it would be applied in practice, but it doesn't make things more complex and would be safer?

```java
if (bucketOrds.size() > targetBuckets) {
  long[] mergeMap;
  do {
    mergeMap = increaseRounding();
  } while (mergeMap.length > targetBuckets);
  // now renumber buckets
}
```